.github/workflows/ci-and-cd-pipeline.yml

```diff
@@ -0,0 +1,43 @@
+name: Continuous Integration and Continuous Delivery
+on: 
+  push:
+    paths-ignore:
+      - 'README.md'
+      - 'documentation/**'      
+  pull_request:
+    paths-ignore:
+      - 'README.md'
+      - 'documentation/**'
+jobs:
+  build:
+    name: Build and test
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - run: docker build --target test --tag todo-app:test .
+    - run: docker run --env-file ./.env.test todo-app:test
+  publish:
+    name: Publish
+    needs: build
+    runs-on: ubuntu-latest
+    if: ${{ (github.ref == 'refs/heads/main' || github.ref == 'refs/heads/module8') }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: sign into Docker
+      env: 
+        DOCKER_HUB_PASSWORD: ${{secrets.DOCKER_HUB_PW}}
+      run: |
+        docker login -u anndevops123 -p $DOCKER_HUB_PASSWORD
```

Comment:
Minor but you might notice that you get a warning from this line: we should ideally action this step using the `--password-stdin` flag rather than passing the password explicitly, to avoid leaking those credentials to shell logs. [See the docs](https://docs.docker.com/engine/reference/commandline/login/) for more info, example of how that might look below:
```
echo $DOCKER_HUB_PASSWORD | docker login -u $anndevops123 --password-stdin
```